File: src/main/java/me/testaccount666/serversystem/clickablesigns/SignManager.java
package me.testaccount666.serversystem.clickablesigns;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class SignManager {
    private final Map<Location, SignType> _signTypes = new HashMap<>();
    private final File _signDirectory = Path.of("plugins", "ServerSystem", "data", "signs").toFile();

    public void addSignType(Location location, SignType signType) {
        _signTypes.put(location, signType);
    }

    public void removeSignType(Location location) {
        _signTypes.remove(location);
    }

    public Optional<SignType> getSignType(Location location) {
        return Optional.ofNullable(_signTypes.get(location));
    }

    public void loadSignTypes() {
        if (!_signDirectory.exists()) return;
        var files = _signDirectory.listFiles();
        if (files == null) return;

        for (var file : files) {
            var fileConfig = YamlConfiguration.loadConfiguration(file);
            var key = fileConfig.getString("Key");
            if (key == null) continue;
            var signTypeOptional = SignType.getSignTypeByKey(key);
            if (signTypeOptional.isEmpty()) continue;
            var signType = signTypeOptional.get();
            var locationString = file.getName().substring(0, file.getName().indexOf("."));
            var locationSplit = locationString.split("_");
            if (locationSplit.length != 4) continue;

            var worldName = locationSplit[0];
            var xString = locationSplit[1];
            var yString = locationSplit[2];
            var zString = locationSplit[3];

            var world = Bukkit.getWorld(worldName);
            if (world == null) continue;
            var x = Integer.parseInt(xString);
            var y = Integer.parseInt(yString);
            var z = Integer.parseInt(zString);
            var location = new Location(world, x, y, z);
            addSignType(location, signType);
        }
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/SignManager.kt
package me.testaccount666.serversystem.clickablesigns

import me.testaccount666.serversystem.ServerSystem
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.file.YamlConfiguration

class SignManager {
    private val _signTypes = HashMap<Location, SignType>()
    private val _signDirectory = ServerSystem.instance.dataPath.resolve("data").resolve("signs").toFile()

    fun addSignType(location: Location, signType: SignType) = let { _signTypes[location] = signType }

    fun removeSignType(location: Location) = let { _signTypes.remove(location) }

    fun getSignType(location: Location) = _signTypes[location]

    fun loadSignTypes() {
        if (!_signDirectory.exists()) return
        val files = _signDirectory.listFiles() ?: return

        for (file in files) {
            val fileConfig = YamlConfiguration.loadConfiguration(file!!)
            val key = fileConfig.getString("Key") ?: continue
            val signType = SignType.getSignTypeByKey(key) ?: continue
            val locationString = file.name.substring(0, file.name.indexOf("."))
            val locationSplit = locationString.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (locationSplit.size != 4) continue

            val worldName = locationSplit[0]
            val xString = locationSplit[1]
            val yString = locationSplit[2]
            val zString = locationSplit[3]

            val world = Bukkit.getWorld(worldName) ?: continue
            val x = xString.toInt()
            val y = yString.toInt()
            val z = zString.toInt()
            val location = Location(world, x.toDouble(), y.toDouble(), z.toDouble())
            addSignType(location, signType)
        }
    }
}
